bigquery/google/cloud/bigquery/_helpers.py

```diff
@@ -29,6 +30,12 @@
 _RFC3339_MICROS_NO_ZULU = "%Y-%m-%dT%H:%M:%S.%f"
 _TIMEONLY_WO_MICROS = "%H:%M:%S"
 _TIMEONLY_W_MICROS = "%H:%M:%S.%f"
+_PROJECT_PREFIX_PATTERN = re.compile(
+    r"""
+    (?P<project_id>\S+\:[^.]+)\.(?P<dataset_id>[^.]+)(?:$|\.(?P<custom_id>[^.]+)$)
+""",
```

Comment:
Are multiple colons in the project name allowed, or should the pattern be updated?

```py
>>> m = _PROJECT_PREFIX_PATTERN.match("example.com::::project_id.dataset_id.table_id")
>>> m is not None
True
>>> m.group("project_id")
'example.com::::project_id'
```

Reply:
That's a valid point, and it's the same thing with other parts, too (e.g. the max length of dataset ID, hyphens int it, etc.).

In an ideal world we would borrow a regex from the backend, and should indeed not try to guess all possible cases ourselves.

Reply:
@tswast Currently we do not check for things that the backend would reject, e.g. the max allowed length of `dataset_id`, dashes instead of underscores in the names, etc.

I know that we prefer backend errors, but would It still be useful to warn the users about these errors early?
(in a separate PR)

Reply:
> In an ideal world we would borrow a regex from the backend, and should indeed not try to guess all possible cases ourselves.

That's too much to keep in sync. Even though identifier restrictions have been very stable, it can still get out of date. For example, what if we start accepting any unicode for dataset / table identifiers to support other languages?

As much as possible, I'd like to keep the client simple, since in all likelihood we'll forget all the places where we do client-side validation

Reply:
OK thanks, resolving this thread then.